tests/test_category.py

```diff
@@ -0,0 +1,116 @@
+        self.assertEqual(serializer_data, response_data)
+
+    def test_create_anonymousUser(self):
+        response = self.client.post('/category/', self.cdict, format='json')
+        self.assertEqual(response.status_code, 403)
+
+    def test_create_empty_name(self):
+        # test category name can't be empty
+        self.assertTrue(self.client.login(**self.user_info))
+        response = self.client.post('/category/', {'name': ''}, format='json')
+        self.assertEqual(response.status_code, 400)
+
+        response = self.client.post('/category/', {'name': None}, format='json')
+        self.assertEqual(response.status_code, 400)
+
+    def test_create_no_parent(self):
+        # test create category with no parent
+        self.assertTrue(self.client.login(**self.user_info))
+        response = self.client.post('/category/', self.cdict, format='json')
+        self.assertEqual(response.status_code, 201)
+        serializer = CategorySerializer(Category.objects.get(owner=self.user, name=self.cname))
+        self.assertEqual(
+            json.loads(json.dumps(serializer.data)),
+            json.loads(response.content))
+
+    def test_create_with_parent(self):
+        self.assertTrue(self.client.login(**self.user_info))
+        # test create category with parent
+        parent = Category.objects.create(owner=self.user, name='papa')
+        response = self.client.post('/category/', {'name': self.cname, 'parent_id': parent.id}, format='json')
+        self.assertEqual(response.status_code, 201)
+        serializer = CategorySerializer(Category.objects.get(owner=self.user, name=self.cname))
+        self.assertEqual(
+            json.loads(json.dumps(serializer.data)),
+            json.loads(response.content))
+
+        # check all fof the fields needed in the results.
+        self.assertListEqual(
+            list(json.loads(response.content).keys()),
+            ['id', 'parent', 'name', 'owner'])
+
+    def test_retrieve(self):
+        pass
+
+    def test_update(self):
+        pass
+
+    def test_delete(self):
+        pass
+
+    def test_model_duplicate(self):
+        # test create duplicate category with no owner
+        Category.objects.create(name='default')
+        self.assertEqual(Category.objects.count(), 1)
+        with self.assertRaises(ValidationError):
+            Category.objects.create(name='default')
+
+        # test create duplicate category with owner
+        Category.objects.create(name='default', owner=self.user)
+        self.assertEqual(Category.objects.count(), 2)
+        with self.assertRaises(ValidationError):
+            Category.objects.create(name='default', owner=self.user)
+
+    def test_model_delete(self):
+        # test create duplicate category with no owner
+        papa = Category.objects.create(name='default')
+        Category.objects.create(name='default_2', parent=papa)
+        papa.delete()
+        self.assertIsNone(Category.objects.first().parent)
```

Comment:
what happens to default_2 in this case? Does it still exist? Do you want it to exist?